membership-attribute-service/app/services/zuora/payment/ZuoraPaymentService.scala

```diff
@@ -113,59 +90,36 @@ class PaymentService(zuoraService: ZuoraSoapService, planMap: Map[ProductRatePla
-      scheduleOpt.map(_.withCreditBalanceApplied(account.creditBalance))
+  ): Future[Option[Bill]] =
+    zuoraService.previewInvoices(subId, numberOfBills).map { previewInvoiceItems =>
+      val maybeBillingSchedule = BillingSchedule.fromPreviewInvoiceItems(planMap, previewInvoiceItems)
+      maybeBillingSchedule.flatMap { billingSched =>
+        billingSched
+          .withCreditBalanceApplied(account.creditBalance)
+          .invoices
+          .list
+          .find(_.amount > 0)
+          .toOption
+      }
     }
-  }
 
-  def billingSchedule(subId: Id, accountId: AccountId, numberOfBills: Int)(implicit logPrefix: LogPrefix): Future[Option[BillingSchedule]] =
-    billingSchedule(subId, zuoraService.getAccount(accountId), numberOfBills)
-
-  def getPaymentMethod(accountId: AccountId, defaultMandateIdIfApplicable: Option[String] = None)(implicit
+  def getPaymentMethod(defaultPaymentMethodId: Option[String], defaultMandateIdIfApplicable: Option[String] = None)(implicit
       logPrefix: LogPrefix,
   ): Future[Option[PaymentMethod]] =
-    getPaymentMethodByAccountId(accountId).map(_.flatMap(buildPaymentMethod(defaultMandateIdIfApplicable)))
-
-  private def getPaymentMethodByAccountId(accountId: AccountId)(implicit logPrefix: LogPrefix): Future[Option[Queries.PaymentMethod]] = {
-    val accountFuture = {
-      val account = zuoraService.getAccount(accountId)
-      account.onComplete {
-        case Failure(exception) =>
-          val message = scrub"Failed to get account for account $accountId"
-          logger.error(message, exception)
-        case Success(_) => logger.info(s"Successfully got account for $accountId")
-      }
-      account
-    }
-    def getPaymentMethod(account: Account) = {
-      val paymentMethod = getPaymentMethodByAccount(account)
-      paymentMethod.onComplete {
-        case Failure(exception) =>
-          val message = scrub"Failed to get payment method for account $accountId"
-          logger.error(message, exception)
-        case Success(_) => logger.info(s"Successfully got payment method for $accountId")
-      }
-      paymentMethod
-    }
     for {
-      account <- accountFuture
-      paymentMethod <- getPaymentMethod(account)
-    } yield paymentMethod
-  }
+      maybePaymentMethod <- getPaymentMethodByAccount(defaultPaymentMethodId).withLogging(s"get payment method for $defaultPaymentMethodId")
+    } yield for {
+      soapPaymentMethod <- maybePaymentMethod
+      memsubPaymentMethod <- buildPaymentMethod(defaultMandateIdIfApplicable, soapPaymentMethod)
+    } yield memsubPaymentMethod
 
-  private def getPaymentMethodByAccount(account: Account)(implicit logPrefix: LogPrefix): Future[Option[Queries.PaymentMethod]] = {
-    val maybeEventualPaymentMethod = account.defaultPaymentMethodId.map(zuoraService.getPaymentMethod)
-    maybeEventualPaymentMethod match {
-      case Some(eventualPaymentMethod) => eventualPaymentMethod.map(Some.apply)
+  private def getPaymentMethodByAccount(
```

Comment:
Is this the right name now? It doesn't look like it does anything related to an account now? Before it took an `Account` as a parameter.

Reply:
good point, it's by the default payment method id (there was no point passing in the whole account to use just one field)